### PR TITLE
Fixes [SR-78] swift compiler seg fault

### PR DESCRIPTION
**Pull request by manavgabhawala**
Fixes [SR-78](https://bugs.swift.org/browse/SR-78) swift compiler segmentation fault. The IRGen module would crash when a generic (empty) struct was part of the source code because it tried to call std::next on an end iterator, adds in a guard to prevent against that and also adds a test case to add in this check.

Added test case that caused the old compiler to crash (I'm not sure if this is the right way to integrate it in the testing suite).

### CODE REVIEW

**Comment by krememek** [comment]
@jckarter: can you review this?

**Comment by simonmpilkington** [comment]
As the return value for this new use case (when the iterator is exhausted) is the same as when there are multiple stored properties, I would handle these in the same code path - more specifically: the only special case here is when there is a single stored property and everything else should use the main code path.

File: test/IRGen/generic_structs.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+struct D<T2>
+{}
+
+struct Foo<A1, A2>
+{
+  var b: Bar<A1, A2>
+}
+
+struct Bar {
+}
```
Comment (blackrain) [inline test/IRGen/generic_structs.swift]: Empty end line is missing.

**Comment by jckarter** [comment]
LGTM. Add the missing newline like @blackrain suggested and I'll accept.

**Comment by manavgabhawala** [comment]
@jckarter I added the newline.

**Comment by jckarter** [comment]
Perfect. Thanks @manavgabhawala!

**Pull request merged**